File: javaclient/src/main/java/confluo/rpc/RPCRecordBatchBuilder.java
package confluo.rpc;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;

/**
 * A builder for a batch of records
 */
public class RPCRecordBatchBuilder {

  /**
   * TIME_BLOCK: The size of a time block for a record batch
   */
  static double TIME_BLOCK = 1e6;

  private long numRecords;
  private TreeMap<Long, ByteArrayOutputStream> batch;
  private Schema schema;

  /**
   * Initializes an empty rpc record batch builder
   */
  public RPCRecordBatchBuilder(Schema schema) {
    this.numRecords = 0;
    this.schema = schema;
  }

  /**
   * Adds a record to the batch builder
   *
   * @param record The record to add to the batch builder
   */
  public void addRecord(ByteBuffer record) throws IOException {
    long ts = record.getLong();
    long timeBlock = (int) (ts / TIME_BLOCK);
    if (!batch.containsKey(timeBlock)) {
      batch.put(timeBlock, new ByteArrayOutputStream());
    }
    batch.get(timeBlock).write(record.array());
    numRecords += 1;

  }

  /**
   * Gets the record batch
   *
   * @return The record batch containing the records
   */
  public rpc_record_batch getBatch() throws IOException {
    rpc_record_batch ret = new rpc_record_batch();
    ret.setNrecords(batch.size());
    for (Map.Entry<Long, ByteArrayOutputStream> entry: batch.entrySet()) {
      long timeBlock = entry.getKey();
      byte[] data = entry.getValue().toByteArray();
      entry.getValue().close();
      long nRecords = data.length / schema.getRecordSize();
      ret.addToBlocks(new rpc_record_block(timeBlock, ByteBuffer.wrap(data), nRecords));
    }
    clear();
    return ret;
  }

  /**
   * Clears the record batch builder
   */
  public void clear() {
    batch.clear();
    numRecords = 0;
  }

}

File: pyclient/confluo/rpc/rpc_record_batch_builder.py
from collections import defaultdict
from ttypes import rpc_record_batch, rpc_record_block
import struct

class rpc_record_batch_builder:
    """
    A builder for a batch of records
    Attributes:
        TIME_BLOCK: The size of a time block for a record batch
    """
    TIME_BLOCK = 1e6
    
    def __init__(self):
        """
        Initializes an empty rpc record batch builder
        """
        self.num_records_ = 0
        self.clear()

    def add_record(self, record):
        """
        Adds a record to the batch builder
        Args:
            record: The record to add to the batch builder
        """
        ts = struct.unpack('l', record[:8])[0]
        time_block = int(ts / self.TIME_BLOCK)
        self.batch_sizes_[time_block] = len(record) + self.batch_sizes_.get(time_block, 0)
        self.batch_[time_block].append(record)
        self.num_records_ += 1

    def get_batch(self):
        """
        Gets the record batch
        Returns:
            The record batch containing the records
        """
        batch = rpc_record_batch([], self.num_records_)
        for time_block in self.batch_:
            data = "".join(self.batch_[time_block])
            num_records = len(self.batch_[time_block])
            batch.blocks.append(rpc_record_block(time_block, data, num_records))
        self.clear()
        return batch

    def clear(self):
        """
        Clears the record batch builder
        """
        self.batch_ = defaultdict(list)
        self.batch_sizes_ = {}
